apps/cdn/src/cache.rs

```diff
@@ -0,0 +1,222 @@
+				.vary_headers
+				.iter()
+				.cloned()
+				.filter_map(|h| headers.remove(&h).map(|v| (h, v)))
+				.collect();
+
+			let cache_key = CacheKey {
+				path: key.clone(),
+				important_headers,
+			};
+
+			if let Some(hit) = self.inner.get(&cache_key).await {
+				hit.hits.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
+
+				return Ok(hit);
+			}
+		}
+
+		// miss
+
+		// request file
+		let response = self
+			.http_client
+			.get(format!("{}/{}", global.config.cdn.bucket_url, key))
+			.timeout(std::time::Duration::from_secs(5))
+			.send()
+			.await?;
+
+		if response
+			.headers()
+			.get(header::VARY)
+			.is_some_and(|v| v.to_str().is_ok_and(|v| v == "*"))
+		{
+			// uncacheable
+			return Ok(CachedResponse::from_reqwest(response).await?);
+		}
+
+		let vary_headers: Vec<_> = response
+			.headers()
+			.get(header::VARY)
+			.and_then(|v| v.to_str().ok())
+			.map(|v| v.split(',').filter_map(|v| HeaderName::from_str(v.trim()).ok()).collect())
+			.unwrap_or_default();
+
+		let cached = CachedResponse::from_reqwest(response).await?;
+
+		let meta = Arc::new(PathMeta { vary_headers });
+
+		self.path_meta
+			.insert(key.clone(), Arc::clone(&meta))
+			.map_err(|_| CacheError::PathMetaUpdate)?;
+
+		let important_headers = meta
+			.vary_headers
+			.iter()
+			.cloned()
+			.filter_map(|h| headers.remove(&h).map(|v| (h, v)))
+			.collect();
+
+		let cache_key = CacheKey {
+			path: key,
+			important_headers,
+		};
+
+		self.inner.insert(cache_key, cached.clone()).await;
+
+		Ok(cached)
+	}
+}
+
+pub struct PathMeta {
+	pub vary_headers: Vec<HeaderName>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct CacheKey {
+	pub path: String,
+	pub important_headers: Vec<(header::HeaderName, header::HeaderValue)>,
+}
+
+#[derive(Debug, Clone)]
+pub struct CachedResponse {
+	pub status: StatusCode,
+	pub date: chrono::DateTime<chrono::Utc>,
+	pub headers: HeaderMap,
+	pub body: bytes::Bytes,
+	pub max_age: std::time::Duration,
+	pub hits: Arc<AtomicUsize>,
+}
+
+impl IntoResponse for CachedResponse {
+	fn into_response(mut self) -> Response<Body> {
+		let hits = self.hits.load(std::sync::atomic::Ordering::Relaxed);
+
+		let age = chrono::Utc::now() - self.date;
+
+		self.headers.insert("x-7tv-cache-hits", hits.to_string().try_into().unwrap());
+		self.headers.insert(
+			"x-7tv-cache",
+			if hits == 0 {
+				HeaderValue::from_static("miss")
+			} else {
+				HeaderValue::from_static("hit")
+			},
+		);
+		self.headers.insert("age", age.num_seconds().to_string().try_into().unwrap());
+
+		(self.status, self.headers, self.body).into_response()
+	}
+}
+
+impl CachedResponse {
+	pub async fn from_reqwest(res: reqwest::Response) -> reqwest::Result<Self> {
+		let cache_control = res
+			.headers()
+			.get(header::CACHE_CONTROL)
+			.and_then(|v| v.to_str().ok())
+			.map(|c| c.to_ascii_lowercase());
+
+		let max_age = cache_control
+			.as_deref()
+			.and_then(|c| c.split(',').find_map(|v| v.strip_prefix("max-age=")))
```

Comment:
This is likely good enough but technically max-age isnt the option we are looking for. https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control

Reply:
Do you think just parsing `max-age` is enough or should we add support for things like `no-store`?

Reply:
Well no technically we are supposed to use `s-maxage` if its present and if not go `max-age`